docs/proposals/stateful-apps.md

```diff
@@ -0,0 +1,392 @@
+first member is partitioned from the cluster, it may not observe the other two members, and thus create its
+own cluster of size 1. The other two members do see the first member, so they form a cluster of size 3.
+Both clusters appear to have quorum, which can lead to data loss if not detected.
+
+StatefulSets should provide basic mechanisms that enable a consistent view of cluster state to be possible,
+and in the future provide more tools to reduce the amount of work necessary to monitor and update that
+state.
+
+The first mechanism is that the StatefulSet controller blocks creation of new pods until all previous pods
+are reporting a healthy status. The StatefulSet controller uses the strong serializability of the underyling
+etcd storage to ensure that it acts on a consistent view of the cluster membership (the pods and their)
+status, and serializes the creation of pods based on the health state of other pods. This simplifies
+reasoning about how to initialize a StatefulSet, but is not sufficient to guarantee split brain does not
+occur.
+
+The second mechanism is having each "member" use the state of the cluster and transform that into cluster
+configuration or decisions about membership. This is currently implemented using a side car container
+that watches the master (via DNS today, although in the future this may be to endpoints directly) to
+receive an ordered history of events, and then applying those safely to the configuration. Note that
+for this to be safe, the history received must be strongly consistent (must be the same order of
+events from all observers) and the config change must be bounded (an old config version may not
+be allowed to exist forever). For now, this is known as a 'babysitter' (working name) and is intended
+to help identify abstractions that can be provided by the StatefulSet controller in the future.
+
+
+## Future Evolution
+
+Criteria for advancing to beta:
+
+* StatefulSets do not accidentally lose data due to cluster design - the pod safety proposal will
+  help ensure StatefulSets can guarantee **at most one** instance of a pod identity is running at
+  any time.
+* A design consensus is reached on StatefulSet upgrades.
+
+Criteria for advancing to GA:
+
+* StatefulSets solve 80% of clustered software configuraton with minimal input from users and are safe from common split brain problems
+  * Several representative examples of StatefulSets from the community have been proven/tested to be "correct" for a variety of partition problems (possibly via Jepsen or similar)
+  * Sufficient testing and soak time has been in place (like for Deployments) to ensure the necessary features are in place.
+* StatefulSets are considered easy to use for deploying clustered software for common cases
+
+Requested features:
+
+* IPs per member for clustered software like Cassandra that cache resolved DNS addresses that can be used outside the cluster
```

Comment:
Does this mean it would be possible to access a specific cluster member from the outside of the cloud? Or by outside of the cluster you mean "within the same namespace"?